src/segment.rs

```diff
@@ -61,77 +53,70 @@ impl Segment {
+            .into_iter()
+            .filter_map(|e| e)
+            .collect::<Vec<ANSIString>>()
     }
+}
 
-    /// Create a string with the formatted contents of a segment
-    ///
-    /// Will recursively also format the prefix and suffix of the segment being
-    /// stringified.
-    pub fn output(&self) -> String {
-        let Segment {
-            name: _name,
-            prefix,
-            value,
-            style,
-            suffix,
-        } = self;
-
-        let mut segment_string = String::new();
-
-        // Skip the prefix for the first segment
-        if let Some(prefix) = prefix {
-            segment_string += &prefix.output()
-        }
+impl fmt::Display for Segment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let ansi_strings = self.ansi_strings();
+        write!(f, "{}", ANSIStrings(&ansi_strings))
+    }
+}
 
-        segment_string += &style.paint(value).to_string();
+/// Segment affixes are to be used for the prefix or suffix of a segment.
+/// By default they will inherit the styling of its segment, unless otherwise specified.
+pub struct SegmentAffix {
+    /// The affix's name, to be used in configuration and logging.
+    name: String,
 
-        if let Some(suffix) = suffix {
-            segment_string += &suffix.output();
-        }
+    /// The affix's style. If None, will inherit the style of the segment containing it.
+    style: Option<Style>,
 
-        segment_string
-    }
+    /// The string value of the affix.
+    value: String,
+}
 
-    /// Create a string with the formatted contents of a segment while skipping the first segment.
-    ///
-    /// Will recursively also format the prefix and suffix of the segment being
-    /// stringified.
-    pub fn output_index(&self, index: usize) -> String {
-        let Segment {
-            name: _name,
-            prefix,
-            value,
-            style,
-            suffix,
-        } = self;
-
-        let mut segment_string = String::new();
-
-        // Skip the prefix for the first segment
-        if index != 0 {
-            if let Some(prefix) = prefix {
-                segment_string += &prefix.output_index(index)
-            }
+impl SegmentAffix {
+    /// Creates a segment affix with no contents.
+    pub fn new() -> SegmentAffix {
+        SegmentAffix {
+            name: String::new(),
+            style: None,
+            value: String::new(),
         }
+    }
 
-        segment_string += &style.paint(value).to_string();
-
-        if let Some(suffix) = suffix {
-            segment_string += &suffix.output();
+    /// Generates the colored ANSIString output.
+    pub fn ansi_string(&self) -> ANSIString {
+        match self.style {
+            Some(style) => style.paint(&self.value),
+            None => ANSIString::from(&self.value),
         }
-
-        segment_string
     }
 }
 
-type BoxedSegment = Option<Box<Segment>>;
+impl fmt::Display for SegmentAffix {
```

Comment:
Could we please have a quick comment here for this section?

Reply:
This particular snippet is mostly boilerplate that is used when defining the `std::fmt::Display` implementation. The part that isn't boilerplate is just `self.ansi_string()`. 😄